jgrapht-core/src/main/java/org/jgrapht/graph/GraphWalk.java

```diff
@@ -174,6 +201,229 @@ public String toString()
         else
             return edgeList.toString();
     }
+
+    @Override
+    public boolean equals(Object o){
+        if(o==null || !(o instanceof GraphWalk))
+            return false;
+        else if(this==o)
+            return true;
+        @SuppressWarnings("unchecked")
+        GraphWalk<V,E> other=(GraphWalk<V,E>) o;
+        if(this.isEmpty() && other.isEmpty())
+            return true;
+        if(!this.startVertex.equals(other.getStartVertex()) || !this.endVertex.equals(other.getEndVertex()))
+            return false;
+        //If this path is expressed as a vertex list, we may get away by comparing the other path's vertex list
+        //This only works if its vertexList identifies a unique path in the graph
+        if(this.edgeList == null && !other.getGraph().getType().isAllowingMultipleEdges())
+            return this.vertexList.equals(other.getVertexList());
+        else //Unlucky, we need to compare the edge lists,
+            return this.getEdgeList().equals(other.getEdgeList());
+    }
+
+    @Override
+    public int hashCode(){
+        if(isEmpty())
+            return 1;
+
+        int hashCode=startVertex.hashCode()+31*endVertex.hashCode();
+        if(edgeList != null)
+            return hashCode+31*edgeList.hashCode();
+        else
+            return hashCode+31*vertexList.hashCode();
+    }
+
+    /**
+     * Reverses the direction of the path. In case of directed/mixed graphs, the arc directions will be reversed.
+     * An exception is thrown if reversing an arc (u,v) is impossible because arc (v,u) is not present in the graph.
+     * @param walkWeightCalculator Function used to calculate the weight of the reversed GraphWalk
+     * @throws InvalidGraphPathException if the walk cannot be reversed.
+     * @return a reversed GraphWalk
+     */
+    public GraphWalk<V,E> reverse(Function<GraphWalk<V,E>,Double> walkWeightCalculator) throws InvalidGraphPathException {
```

Comment:
Good idea; i'll add it. I added the weight calculator because I frequently encounter problems where the weights are on the vertices and not on the arcs (weight of the path is sum of weights of the vertices), so having some more flexibility in the definition of the weight of a path is nice.